File: leetcode/trim-a-binary-search-tree/tree-and-recursion.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func trimBST(root *TreeNode, L int, R int) *TreeNode {
	if nil == root {
		return nil
	}

	left, right := trimBST(root.Left, L, R), trimBST(root.Right, L, R)
	if root.Val > R {
		return left
	} else if root.Val < L {
		return right
	} else {
		root.Left, root.Right = left, right
		return root
	}
}

File: leetcode/lowest-common-ancestor-of-a-binary-tree/tree-and-recursion.cpp
#include <iostream>
#include <string>
#include <stack>
#include <queue>
#include <deque>
#include <vector>
#include <assert.h>
#include <unordered_map>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if (NULL == p || NULL == q){
      return NULL;
    }
    if (NULL == root || p == root || q == root || p == q){
      return root;
    }

    TreeNode *left = lowestCommonAncestor(root->left, p, q);
    TreeNode *right = lowestCommonAncestor(root->right, p, q);

    if (NULL == left){
      return right;

    } else if (NULL == right){
      return left;

    } else {
      return root;
    }
  }
};
